benchmark/benchmark-atomic.cpp

```diff
@@ -0,0 +1,220 @@
+/// RAJA's implementations.  Wrapping the call in an functor allows simple type deduction for printing
+/// from within the benchmark.
+template<typename T, typename DesulAtomicSignature<T>::type atomic_impl>
+struct atomicWrapperDesul {
+    /// Call operator overload template that allows invoking DESUL atomic with a (int*)(T*, T) signature
+    RAJA_DEVICE T operator()(T * acc, T value) const {
+        return desul::atomic_fetch_add(acc, value, raja_default_desul_order{},
+                        raja_default_desul_scope{});
+    }
+};
+
+template<typename T>
+class IsDesul : public std::false_type {};
+
+template<typename T, typename DesulAtomicSignature<T>::type atomic_impl>
+class IsDesul<atomicWrapperDesul<T, atomic_impl>> : public std::true_type {};
+
+template<typename AtomicImplType>
+std::string GetImplName (const AtomicImplType& impl) {
+    if (IsDesul<AtomicImplType>::value) {
+        return "Desul atomic";
+    } else {
+        return "RAJA atomic";
+    }
+}
+
+template <class ExecPolicy, typename AtomicType, bool test_array, typename AtomicImplType>
+void TimeAtomicOp( AtomicImplType atomic_impl, uint64_t N, uint64_t num_iterations = 4, int array_size = 100, bool print_to_output = true) {
+    RAJA::Timer timer;
+
+    // Allocate memory
+    AtomicType* device_value = nullptr;
+    int len_array = test_array ? array_size : 1;
+    camp::resources::Resource resource {RAJA::resources::get_resource<typename ExecPolicy::policy>::type::get_default()};
+    device_value = resource.allocate<AtomicType>(len_array);
+
+    timer.start();
+    for (int i = 0; i < num_iterations; ++i) {
+        RAJA::forall<typename ExecPolicy::policy>(RAJA::RangeSegment(0, N),
+        [=] RAJA_HOST_DEVICE(int tid)  {
+            if (test_array) {
+                atomic_impl(&(device_value[tid % array_size]), AtomicType(1));
+            } else {
+                atomic_impl(device_value, AtomicType(1));
+            }
+        });
+    }
+
+    resource.wait();
+    timer.stop();
+    resource.deallocate(device_value);
+
+    double t = timer.elapsed();
+    if (print_to_output) {
+        std::cout << INDENT << INDENT << t << "s" << INDENT;
+        std::cout << GetImplName(atomic_impl) << ", ";
+        std::cout << "Number of atomics under contention " << array_size << ", ";
+        std::cout << num_iterations * N << " many atomic operations" << ", ";
+        std::cout << ExecPolicy::PolicyName();
+        std::cout << std::endl;
+    }
+}
+
+int main (int argc, char* argv[]) {
+    if (argc > 2) {
+        RAJA_ABORT_OR_THROW("Usage: ./benchmark-atomic.exe <N> where N is the optional size of the benchmark loop");
+    }
+    uint64_t N = 1000000000;
+    if (argc == 2) {
+        N = std::stoi(argv[1]);
+    }
+
+    // Perform an untimed initialization of both desul and RAJA atomics.
+    TimeAtomicOp<ExecPolicyGPU<BLOCK_SZ>, int, true>(AtomicAdd<typename GPUAtomic::policy>{}, N, 10, 1000, false);
```

Comment:
How are the numerical values passed to the method that runs the loop determined?